filebeat/reader/readfile/decoder.go

```diff
@@ -0,0 +1,147 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package readfile
+
+import (
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+// decoderReader reads from an input file and converts its content into UTF-8.
+// It stores the size of the original bytes, so offset can be calculated later in
+// the line scanner.
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	offset     int // segment byte offset
+
+	bufferSize int
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) (*decoderReader, error) {
+	r := &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		offset:     0, // TODO init correctly
+		bufferSize: bufferSize,
+	}
+
+	return r, nil
+}
+
+// seekToLastRead reads from the input file until the saved offset is reached.
+// It is the substitution of seeking in the file.
+// This function is no longer required after registry refactoring is done.
+func (r *decoderReader) seekToLastRead() error {
+	if r.offset == 0 {
+		return nil
+	}
+
+	remaining := r.offset
+	for remaining > 0 {
+		s := r.bufferSize
+		if remaining < s {
+			s = remaining
+		}
+
+		b := make([]byte, s)
+		n, err := r.in.Read(b)
+		remaining -= n
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+// read reads from an underlying file and converts it into UTF-8.
+func (r *decoderReader) read(buf []byte) (int, error) {
+	// collect leftover converted bytes
+	nb := 0
+	var err error
+	if r.buf.Len() != 0 {
+		nb, err = r.copyToOut(buf)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	b := make([]byte, r.bufferSize)
+	start := r.encodedBuf.Len()
+	// read from the underlying file
+	n, err := r.in.Read(b[start:])
+	if err != nil {
+		if err == io.EOF && start == 0 {
+			return 0, err
+		}
+	}
+
+	// collect leftover encoded bytes
+	if start > 0 {
+		enc, err := r.encodedBuf.Collect(start)
+		if err != nil {
+			return 0, err
+		}
+		r.encodedBuf.Reset()
+		b = append(enc, b[start:]...)
+	}
+
+	// convert encoded bytes into UTF-8
+	nDst, nSrc, err := r.decoder.Transform(buf[nb:], b[:start+n], false)
+	if err != nil {
+		if err == transform.ErrShortSrc {
+			r.encodedBuf.Append(b[nSrc:])
+			r.offset += nSrc
+			return nDst, nil
+		}
+		if err == transform.ErrShortDst {
+			r.encodedBuf.Append(b[nSrc : start+n])
+			r.offset += nSrc
+			return nDst, nil
+		}
+		return 0, err
+	}
+
+	r.offset += nSrc
+	return nDst, nil
+}
```

Comment:
this function + decoderReader definition are super close to the original code. I think we can simplify the code remove one array copy per read by just throwing away the old code and write the function from scratch. No need for `buf` and `encodedBuf`. We can directly decode into the input buffer.

Can we make use [Decoder.Reader](https://godoc.org/golang.org/x/text/encoding#Decoder.Reader) directly? Decoder reader just uses [transform.Reader](https://godoc.org/golang.org/x/text/transform#Reader).

Reply:
I wanted to use it originally, but `Read` returns only the length of the converted string. To track both the offset of the encoded file and the converted stream I needed `Transform` which returns how many bytes it has consumed and how many it has produced.